File: solutions/javascript/clock/1/clock.js
// For my Shree DR.MDD

export class Clock {
  constructor(hours = 0, minutes = 0) {
    this._normalizeTime(hours, minutes);
  }

  _normalizeTime(hours = 0, minutes = 0) {
    let totalMinutes = (hours * 60 + minutes) % (24 * 60);
    if (totalMinutes < 0) {
      totalMinutes += 24 * 60;
    }
    this._hour = Math.floor(totalMinutes / 60);
    this._minute = totalMinutes % 60;
  }

  toString() {
    return `${this._hour < 10 ? '0' : ''}${this._hour}:${this._minute < 10 ? '0' : ''}${this._minute}`;
  }

  plus(minutes) {
    this._normalizeTime(this._hour, this._minute + minutes);
    return this;
  }

  minus(minutes) {
    this._normalizeTime(this._hour, this._minute - minutes);
    return this;
  }

  equals(other) {
    if (!(other instanceof Clock)) {
      return false;
    }
    return this._hour === other._hour && this._minute === other._minute;
  }
}

File: solutions/lua/clock/1/clock.lua
local Clock = {}
Clock.__index = Clock

local MINUTES_IN_DAY = 24 * 60
    
function Clock.at(hours, minutes)
    return Clock:new(hours * 60 + (minutes or 0))
end
    
function Clock:new(minutes)
    while minutes < 0 do minutes = minutes + MINUTES_IN_DAY end
    while minutes >= MINUTES_IN_DAY do minutes = minutes - MINUTES_IN_DAY end
    return setmetatable({ minutes = minutes }, self)
end
function Clock:plus(minutes) return Clock:new(self.minutes + minutes) end
function Clock:minus(minutes) return self:plus(-minutes) end
function Clock:equals(clock) return self.minutes == clock.minutes end
function Clock:__tostring()
    return string.format("%02d:%02d", math.floor(self.minutes / 60), self.minutes % 60)
end
        
return Clock